ops/charm.py

```diff
@@ -1417,6 +1423,103 @@ def config(self) -> model.ConfigData:
+        **kwargs: Any,
+    ) -> _ConfigType:
+        """Load the config into an instance of a config class.
+
+        The raw Juju config is passed to the config class's ``__init__``, as
+        keyword arguments, with the following changes:
+
+        * ``secret`` type options have a :class:`model.Secret` value rather
+          than the secret ID. Note that the secret object is not validated by
+          Juju at this time, so may raise :class:`SecretNotFoundError` when it
+          is later used if the secret does not exist or the unit does not have
+          permission to access it.
+        * dashes in names converted to underscores.
+
+        Any additional positional or keyword arguments to this method will be
+        passed through to the config class ``__init__``.
+
+        Args:
+            cls: A class that will accept the Juju options as keyword arguments.
+            errors: what to do if the config is invalid. If ``blocked``, the
+                charm will exit successfully (this informs Juju that
+                the event was handled and it will not be retried) after setting
+                an appropriate blocked status. If ``raise``, ``load_config``
+                will not catch any exceptions, leaving the charm to handle
+                errors.
+            args: positional arguments to pass through to the config class.
+            kwargs: keyword arguments to pass through to the config class.
+
+        Returns:
+            An instance of the config class with the current config values.
+
+        Raises:
+            ValueError: if the configuration is invalid and ``errors`` is set to
+                ``raise``.
+        """
+        from ._main import _Abort
+
+        config: dict[str, bool | int | float | str | model.Secret] = kwargs.copy()
+        fields = set(_juju_option_names(cls))
+        for key, value in self.config.items():
+            attr = key.replace('-', '_')
+            if not attr.isidentifier():
+                if errors == 'raise':
+                    raise ValueError(f'Invalid attribute name {attr}')
+                self.unit.status = model.BlockedStatus(f'Invalid attribute name {attr}')
+                raise _Abort(0)
+            if attr not in fields:
+                continue
+            option_type = self.meta.config.get(key)
+            # Convert secret IDs to secret objects. We create the object rather
+            # that using model.get_secret so that it's entirely lazy, in the
+            # same way that SecretEvent.secret is.
+            if option_type and option_type.type == 'secret':
+                assert isinstance(value, str)  # Juju will have made sure of this.
+                value = model.Secret(
+                    backend=self.model._backend,
+                    id=value,
+                    _secret_set_cache=self.model._cache._secret_set_cache,
+                )
+            config[attr] = value
```

Comment:
How will a secret field be type-annotated in the config class? As `ops.Secret`? Is looking it up in meta better than relying on that fact?

Reply:
Presumably as `ops.Secret`, yes. If we look it up in meta then we know that Juju has done some basic validation on the secret, and I think that's better. If we base it on the type annotation of the attribute then people could do things like have a Juju type of string and as long as it was in the form of a secret, it'd still work. I think that's too cutesy/hack-y and we should force people to use the real Juju secret type.